forch/ports_testing_states_manager.py

```diff
@@ -0,0 +1,111 @@
+"""Module that manages the testing states of the access ports"""
+
+import logging
+import threading
+
+from forch.proto.shared_constants_pb2 import TestingState
+
+
+LOGGER = logging.getLogger('portsm')
+STATE_HANDLERS = {}
+
+
+def _register_state_handler(state_name):
+    def register(func):
+        STATE_HANDLERS[state_name] = func
+        return func
+    return register
+
+
+class PortTestingStateMachine:
+    """State machine class that manages testing states of an access port"""
+
+    AUTHENTICATED = 'authenticated'
+    SEQUESTERED = 'sequestered'
+    OPERATIONAL = 'operational'
+    INFRACTED = 'infracted'
+
+    TRANSITIONS = {
+        AUTHENTICATED: {
+            TestingState.cleared: OPERATIONAL,
+            TestingState.sequestered: SEQUESTERED,
+        },
+        SEQUESTERED: {
+            TestingState.passed: OPERATIONAL,
+            TestingState.failed: INFRACTED,
+        },
+    }
+
+    def __init__(self, mac, initial_state):
+        self._mac = mac
+        self._current_state = initial_state
+
+    def handle_testing_state_event(self, testing_state):
+        """Handle testing state event"""
+        to_state = self.TRANSITIONS.get(self._current_state, {}).get(testing_state, {})
+
+        if not to_state:
+            LOGGER.warning(
+                'Cannot find next state for device %s in state %s for testing_state %s',
+                self._mac, self._current_state, testing_state)
+            return
+
+        LOGGER.info(
+            'Device %s is entering %s state from %s state',
+            self._mac, to_state, self._current_state)
+
+        self._current_state = to_state
+        self._handle_current_state()
+
+    def get_current_state(self):
+        """Get current state of the port"""
+        return self._current_state
+
+    def _handle_current_state(self):
+        if self._current_state in STATE_HANDLERS:
+            STATE_HANDLERS[self._current_state](self)
+
+    @_register_state_handler(state_name=AUTHENTICATED)
+    def _handle_authenticated_state(self):
+        LOGGER.info('Handling authenticated state for device %s', self._mac)
+
+    @_register_state_handler(state_name=SEQUESTERED)
+    def _handle_sequestered_state(self):
+        LOGGER.info('Handling sequestered state for device %s', self._mac)
+
+    @_register_state_handler(state_name=OPERATIONAL)
+    def _handle_operational_state(self):
+        LOGGER.info('Handling operational state for device %s', self._mac)
+
+
+class PortsTestingStatesManager:
+    """Manages the testing states of the access ports"""
+    def __init__(self):
+        self._state_machines = {}
+        self._static_testing_states = {}
+        self._lock = threading.Lock()
```

Comment:
Lock granularity as per discussion.

Reply:
I think there are two levels of lock needed here:

1. Lock for the port manger's `_state_machines`: in handle_authenticated_device(), it may create multiple new state machines and insert them into the map. Thus I think we need a lock to avoid concurrent insertion in port manager itself...

2. Lock for each state machine: this could be added per state machine. However, since there is already a lock in port manager, I think we can reuse it when modiying the state machine from port manager...

Reply:
So for 2, I think I will move the lock to state machine level as discussed.